Add support for path completion

Philipp-M opened the pull request:
This PR adds support for path completion.

It supports the following:

* Autocompletion is triggered with `/`.
* Documentation preview (file permissions, canonicalized path).
* shell expansion (e.g. `~/path`, `$HOME/path`, `${HOME}/path`)
* Async (via `spawn_blocking` instead of tokios file accessor functions, as they IMHO make the code less readable and are quite a bit slower than just spawning a "thread")
* Configurable with `editor.path-completion` (default `true`), per-language overrideable path-completion support
* Refactors `goto_file` with a new more robust path-detection/resolution mechanism and also adds shell-expansion.

A baby-step towards #1015

kirawi commented:
I think it would be better to split off the changes to the LSP into a separate PR. Unless completion is dependent upon them?

Philipp-M commented:
You mean all the commits that are also in #2507?

In case, I will rebase this as soon as the relevant commits this PR is dependent on are on master, currently only the last commit is relevant for this PR or different to #2507.

nrdxp commented:
What is left to move this out of draft status? I've been using this for over a week now and it seems to work fine.

Philipp-M commented:
Well for one (biggest blocker): This is currently dependent on a few commits (particularly the extension/merging of the completion menu) of #2507, and I'm unsure how to progress as I'm awaiting some feedback there.

Also support for windows paths should be added, this should be simple (by extending the path regex), but I would like to first resolve the first issue.

Another (smaller, maybe not worth to fix) issue is, that currently only one path per line is possible, this could probably be tackled with a different path-regex. But I've read that every character but `\0` is allowed in a unix filepath, so I think it's difficult to find a good regex that doesn't have (realworld) limitations.

I've also experimented with async io (via `tokio::fs`), but IMHO the performance-drop and probably race-conditions (in the completion menu, if writing to fast/slow IO) wasn't worth it to continue that path (yet).

cole-h commented on helix-term/src/commands.rs:
```diff
@@ -3942,43 +3961,188 @@ pub fn completion(cx: &mut Context) {
+                let items = ui::PATH_REGEX
+                    .find(&line_until_cursor)
+                    .and_then(|matched_path| {
+                        let matched_path = matched_path.as_str();
```
Would it be possible to also add a denylist setting? I run NixOS, so attempting to index `/nix/store` would be extremely slow with its ~265000 files / directories.

Philipp-M replied:
Maybe we can just use the  `.(git)ignore` for that?

Actually I've used `nix/store`  for stress testing/general feeling, and I thought it was acceptable given the amount of files (but I've got a fast CPU and NVME drive for /nix ...))

Would someone actually edit something directly in `/nix/store` or in a directory with that much files/folders?

I'm not sure of a daily usecase currently where this would really be a problem, but I'm open for different solutions (my suggestion would be a global .ignore)

cole-h replied:
I don't _edit_ things in the Nix store, no, but I do use Helix to read files in there very frequently. I'd rather not use a .ignore for that because so many other tools read that, and I'd like to set this only in Helix.

Philipp-M replied:
Right now it's less of a blocker (literally :)) for typing, as it's now async and results will be discarded when not relevant anymore (i.e. the user has typed something that doesn't match the completion item).

On a fast computer it's still *acceptable* IMHO, especially if the folder was indexed once (i.e. is likely cached in memory when accessing again)).

If this is still relevant, can you/someone maybe think/prototype how the UX/configuration for this look like (I'm happy to implement it, but I'm not sure how the UI for this should look like).

But the (completion-)menu is quite slow with > ~100k entries that are sorted synchronously every keystroke.
I think a more sophisticated lazier/async/"streamable" menu/picker implementation would be probably help in general (I still think that the behavior is so similar (also in the future), that the implementations should be shared, but that's a different topic...).

nrdxp commented:
Did something change? After rebasing on the last push path completion seems to have just stopped working all together. Do I need to configure something differently?

Philipp-M commented:
Have you tried using the HEAD of this branch?

My personal fork (rebased on this branch) is still working.

It's a little bit out of sync with master, I will try to find time soon to rebase this and the multiple-language-servers PR onto master.

nrdxp commented:
Oops, I figured it out. It was my own fault, I forgot to push my latest changes of my forked branch (based on this branch) to the remote so my system never pulled my rebase. Apologies :pray:

danielgomez3 commented:
Any status on when this might be merged to main?

Philipp-M commented:
This is not yet ready to merge and needs a major rebase/refactor and support for windows. I'll wait for this at least until multiple language servers support is merged.
I'm also not completely sure if it makes sense to put this into core (could be a super simple language server, that can be enabled for all languages).

nrdxp commented:
Now that multi-language support is merged we should either close this or rebase and merge. I would vote for the later since I think a consistent path completion baked right into the editor makes sense.

Philipp-M commented:
I have now rebased (or more like rewritten) it onto master.

It now uses the existing `lsp::CompletionItem` and adds a new field enum field `source` to `ui::CompletionItem` (that might be extended when more [completion sources](https://github.com/helix-editor/helix/issues/1015) will be implemented, instead of previously a custom type.
I think this is a little bit more portable (could even be used with little modification for a dedicated language server) and requires fewer changes in `completion.rs`. It's generally a cleaner and improved implementation, and should be easier to maintain/review IMHO.

Following is new:
* documentation preview (file type, file permissions, full canonicalized path).
* home-path resolution (`~/path`)
* async (via `spawn_blocking` instead of tokios file accessor, as they IMHO make the code less readable and are quite a bit slower than just spawning a "thread")
* per-language overridable path-completion support

 I think it's now in a good state for review/merging for unix platforms at least, but windows support is still missing, I'm not sure if it should be merged without windows path support...?

I don't have access to a windows machine currently and would like to avoid setting up a VM, maybe someone else can help out here, via a PR to my branch `path-completion`? Maybe *if* I find time for it (I have a few other ongoing projects, and it's not a high priority for me), I'll write some integration tests for windows paths (and for unix...). Help/support here is also appreciated.

Philipp-M commented:
The tests for OSX failed previously (now they run through, pushed a fix), due to `libc::mode_t` being 16 bit on OSX (is this correct?). Since this is only tested on linux, maybe someone with OSX should test this (which I also don't have access to currently). This is relevant for the file permissions in the documentation window, which should look something like `[rwxr-xr--]`.

inalone commented:
Would love for this to be merged, this is the main feature I'm missing from my neovim config.

nrdxp commented:
just an FYI, since this is the only thing stopping me from tracking the latest master, I went ahead and worked out the merge conflict in this branch for now:
https://github.com/nrdxp/helix/tree/fork3

Feel free to steal it to quickly fix the conflicts in this PR branch

Philipp-M commented:
Thanks for doing this! I have rebased this on master as well (merge conflicts are fortunately maintainable).

Well as said, feedback for OSX especially would be appreciated. Whether Windows needs to be supported within this PR as well (but my feeling is, that it should be, but I'm not super keen in adding it, I'd appreciate a PR on top of this PR).

One thing I noticed after implementing this, is that there's similar code already within `helix_term::ui::completers` it's slightly different though, so it may or may not be worth it to share the code between (in-editor-view) completion sources...

Advantage for the current code is, that it's quite portable though.

FWIW I'm maintaining my own fork with the new event-system, rainbow-treesitter matches, and a few other smaller fixes (possibly necessary due to merges) in https://github.com/Philipp-M/helix/tree/personal/

nilsherzig commented:
Nice work :) 

Do you think there would be any problems if the autocompletion would add a '/' at the end of a directory name? This would enable a much smoother "autocompletion chain" for files deep inside some nested folders.

eliliam commented:
Once we get that linting error fixed, is this good to go? I'd love to see this merged in since it seems like a really great implementation of path autocomplete!

sn0n commented:
any updates? I think this is the functionality I'm looking for,.. new to helix amd figure out gf for goto file, but cant figure out the 'insert file name here from this list' type path completion. sorry to bother.

daedroza commented:
@Philipp-M : Hey, can this PR be updated with newest HEAD? The place where we hook isn't commands anymore but in separate completion handler.

Philipp-M commented:
Rebased onto master.

> if the autocompletion would add a '/' at the end of a directory name?

I guess that could be done, possibly configurable, I like the quick retrigger via `/` though (while not triggering a new directory while selecting the item in the menu). Not sure about it...

> any updates?

Well as described in my last comment I'm awaiting general feedback.

nrdxp commented:
I don't think we have proper visibility on this PR, and poor Phillip has been force pushing updates for a few years now. I've been using this patch from the close to the beginning and haven't had any issues with it. Sorry if the ping is inappropriate but @archseer or @the-mikedavis do you guys have an idea of what is missing or why this hasn't been merged in so long?

Or has it just been overlooked?

lukeflo commented:
Thanks for the great work. I just forked the main Helix repo and rebased your PR as well as another PR (#11164) on my forks master. Just built it from there and works great. Was one of the major things which hindered me switching completely to Helix.

Hope it will find its way into the main repo soon.

One thing I encountered is that expanding Tilde (`~`) for home dir works fine, but the environment variable `$HOME` isn't expanded at all. Is there a reason for that or am I missing something concerning the concrete use-case?

eliliam reviewed (approved):
Looks great to me, who are the main people we need to review this to get this finally merged in?

the-mikedavis commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +299,223 @@ fn request_completion(
+    static PATH_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"((?:~)?(?:\.{0,2}/)+.*)$").unwrap());
+
+    let cur_line = text.char_to_line(cursor);
+    let begin_line = text.line_to_char(cur_line);
```
I wonder if we should cap this to some reasonable maximum (maybe some N hundred characters) to prevent degrading the search if you have very long single lines.

Philipp-M replied:
Good point, I made the limit now `-1000` compared to the current cursor.

the-mikedavis commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +299,223 @@ fn request_completion(
+
+    let cur_line = text.char_to_line(cursor);
+    let begin_line = text.line_to_char(cur_line);
+    let line_until_cursor = text
+        .slice(begin_line..cursor)
+        .as_str()
+        .unwrap_or_default()
+        .to_owned();
```
With #9422 it should be possible to operate on the ropeslice directly now. The regex uses `$` though which doesn't work the same way with regex-cursor so it may not be possible to use it yet - regex-cursor and upstream regex might need the `\z` capture to detect the end of the slice

Philipp-M replied:
Thanks for the suggestion. I'm using now the `regex-cursor`, as you say it doesn't support `$` the same way as `regex`, so I worked around that. But since the path detection is not perfect anyway (difficult, since everything apart of `\0` is allowed in unix paths). I think this is good enough for now.

the-mikedavis commented on book/src/generated/typable-cmd.md:
```diff
@@ -2,7 +2,7 @@
 | `:quit`, `:q` | Close the current view. |
 | `:quit!`, `:q!` | Force close the current view, ignoring unsaved changes. |
-| `:open`, `:o` | Open a file from disk into the current view. |
+| `:open`, `:o`, `:edit`, `:e` | Open a file from disk into the current view. |
```
Looks like this snuck in from a merge/rebase

the-mikedavis commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +299,223 @@ fn request_completion(
+                    }
+                }
+                let ends_with_slash = match matched_path.chars().last() {
+                    Some('/') => true, // TODO support Windows
```
Can this be covered with just using the `std::path::MAIN_SEPARATOR`?

the-mikedavis commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +299,223 @@ fn request_completion(
+
+                let edit_diff = typed_file_name.as_ref().map(|f| f.len()).unwrap_or(0);
+
+                let text_edit = Some(lsp::CompletionTextEdit::Edit(lsp::TextEdit {
+                    range: range_to_lsp_range(
+                        &text,
+                        Range::new(cursor - edit_diff, cursor),
+                        OffsetEncoding::default(),
+                    ),
+                    new_text: file_name.clone(),
+                }));
```
I'd like to try to find a way to avoid digging ourselves into lsp-types here. It makes for a pretty small+focused change but I think we should be introducing core types like `helix_core::Diagnostic` where possible or at least wrappers/replacements like #11486. I suspect it's a potentially large refactor but I think this is the right feature to introduce it with since the actual searching/completion part of this is quite small. (Compared to completing words from the buffer for example)

the-mikedavis replied:
I haven't looked into this much but it might be pretty simple to update the `CompletionItem` struct to have an enum between the `lsp::CompletionItem` and a `helix_core::CompletionItem` struct that has a changeset or transaction instead of text_edits

Philipp-M replied:
Actually a previous iteration of this PR (e.g. [this](https://github.com/helix-editor/helix/tree/12c088918405b2debd04d259293489f7a1fc539d)) had an enum `CompletionItem` with separating variants `LSP` and `Path` .
I wasn't sure having a separate mechanism additional to `lsp::CompletionItem` would make sense, since there was already the variants `lsp::CompletionItemKind::{FILE, FOLDER}`.

But I guess when thinking about more completion sources it may make sense, especially when the move to more core types resembling those in `lsp` is planned anyways.

So when I'm getting you right, you mean a `CompletionItem` at a lower-level than in `ui::completion` (which I think was the previous iteration). I wonder if it makes sense to generally abstract over `lsp::CompletionItem` and wrap it in `helix_core::CompletionItem`. I should probably investigate this, as I think you're right in that this would need a larger refactor (*maybe* worth it).

I think a `Transaction` would make more sense here, being more general (and we have `impl From<ChangeSet> for Transaction` already)?

Philipp-M replied:
Ok, I don't think abstracting over `lsp::CompletionItem` makes sense with the current architecture. So rather having a "data" container `helix_core::CompletionItem` similar to `helix_core::Diagnostic` a little bit independent of `lsp::CompletionItem` with data like documentation (possibly lazy) and `Transaction` etc. makes probably more sense for now.

Is it planned btw. to move more logic of this kind from `helix-term` to e.g. `helix-view` or even `helix-core`?
I could imagine `helix_term::ui::CompletionItem` would make more sense in `helix-view`?

the-mikedavis replied:
Yeah generally the more we can move from -term to -view the merrier. Ideally most of what lives in -term right now would be in -view (even commands) but it will take quite a bit of work to decouple the UI elements. So -term would just be the terminal backend and then there could be a -gui crate that would also implement the UI components.

Philipp-M replied:
Ok, good to know, I'm actually thinking about starting that refactor in this PR, that would obviously blow up this PR a little bit, and would likely produce merge-conflicts with other PRs (e.g. #9996), which is why I'm not sure about it, as I'd like to avoid resolving merge-conflicts and/or maintain this PR for a longer time (in a state where this is more interconnected with the rest of the editor-logic).

Like putting a lot of logic of `helix_term::ui::completion` into e.g. `helix-view::completion` as I don't see many dependencies to `helix-term` currently. (Same for the handlers I guess).

But maybe keep it more incremental, and leave the `CompletionItem` (and the path/lsp `CompletionItem` resolution logic) currently in `helix-term`, i.e. only `helix_term::ui::completion::CompletionItem` to an enum roughly like this:

```rust
pub enum PathKind { Folder, File, Link } // maybe not worth it...

pub enum CompletionItem {
    Lsp {
        item: lsp::CompletionItem,
        id: LanguageServerId,
        resolved: bool,
    },
    Path { // Not sure if this should be a more generic `Core`,
           // but this probably allows being more flexible in the future
        kind: PathKind,
        item: helix_core::CompletionItem,
    },
}

pub struct helix_core::CompletionItem {
    pub transaction: Transaction,
    pub label: String,
    /// Containing Markdown, this could be turned e.g. into an enum to be more structured
    pub documentation: String,
}
```

the-mikedavis replied:
Hmm yeah, working incrementally is probably the right way to go about this. Moving more things to helix-view would mainly help the GUI effort which I don't believe anyone is working on actively anyways. I'm also worried about conflicts with #9801 (though on second look it doesn't look like it would conflict much). An enum like that looks good even if it's only a stepping stone to a bigger refactor later.

Philipp-M replied:
Actually I have merged the snipped system with path-completion in my personal branch, I don't remember that this caused any serious merge-conflicts...

Ok, good to know, then I'll go the incremental route (when I find the time).

the-mikedavis commented on helix-term/src/handlers/completion.rs:
```diff
@@ -346,7 +571,10 @@ pub fn trigger_auto_completion(
         });
-    if is_trigger_char {
+
+    let trigger_path_completion = text.ends_with("/") && doc.supports_path_completion();
```
Same here with maybe being able to use `std::path::MAIN_SEPARATOR`

the-mikedavis commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +299,223 @@ fn request_completion(
+                            ]
+                            .iter()
+                            .fold(String::new(), |mut acc, (p, s)| {
+                                #[allow(clippy::unnecessary_cast)]
+                                acc.push(if mode & (*p as u32) > 0 { *s } else { '-' });
```
```suggestion
                                acc.push(if mode & p > 0 { *s } else { '-' });
```

it looks like the compiler derefs this automatically now

the-mikedavis replied:
Actually I guess you could use `into_iter` instead of iter here and also drop the deref for `s`

Philipp-M replied:
Yeah `into_iter` is possible, also added `String::with_capacity(9)` to avoid unnecessary allocs...

Btw. the cast is necessary for e.g. OSX (added a comment as well) as there is `p: u16`.

Philipp-M commented:
Ok, did a bit of refactoring to the `CompletionItem::{Lsp, Path}` enum. I tried to keep the diff small, but it obviously grew a little bit, since the additional variant needs a little bit different handling.

I've added the basically unchanged
```rust
struct helix_core::CompletionItem {
    pub transaction: Transaction,
    pub label: String,
    /// Containing Markdown
    pub documentation: String,
}
```

I think that could be extended over time when necessary.

> One thing I encountered is that expanding Tilde (`~`) for home dir works fine, but the environment variable `$HOME` isn't expanded (means recognized as e.g. `/home/user`) at all. Is there a reason for that or am I missing something concerning the concrete use-case?

I've also added support for this along the way.

Philipp-M commented:
Hmm nightly rustc seems to be able to handle a few more borrow-checking cases it seems, something like:

```rust
let transaction = match item {
    Arm1 { item, .. } => &item_to_transaction(item),
    Arm2 { item, .. } => &item.transaction
};
```
doesn't seem to be possible without nightly...

Philipp-M commented:
Btw. I have also refactored the handling of additional text edits of lsp completion, composing it now as a single `Transaction` (as it was a simple change, and there was a `TODO` comment anyway for this).

Philipp-M commented:
> Btw. I have also refactored the handling of additional text edits of lsp completion, composing it now as a single `Transaction`

Well I just reverted that change again... I thought compose handles different resulting `ChangeSet::len`s, which is not the case. So that would be more complicated, and should probably be handled in a different PR.

the-mikedavis commented:
Interesting that nightly allows borrowing that value. An alternative could be wrapping it in a `Cow::Owned` and the path variant uses `Cow::Borrowed` but passing it to `Document::apply_temporary` in both branches is good too.

the-mikedavis commented on helix-term/src/handlers/completion/resolve.rs:
```diff
@@ -26,6 +27,15 @@ pub struct ResolveHandler {
     resolver: Sender<ResolveRequest>,
 }
 
+macro_rules! lsp_variant {
```
I'd like to avoid the macro here if possible - I think we can be explicit in the places where we want to match the `CompletionItem::Lsp` variant instead. For `ensure_item_resolved` that means we need to move the `if self.last_request.as_deref()...` check to the top of the function because of the mutable borrow of `item` and then we can say

```rust
let CompletionItem::Lsp(lsp_item) = item else {
    return;
}
```

and we can unconditionally pass the `item` to `ensure_item_resolved` in `helix-term/src/ui/completion.rs`. In the `ResolveRequest::execute` function we can do basically the same but use `unreachable!` (the same as what you have now but inlining the macro)

Philipp-M replied:
Before I'm repeating work again, do you mean something like [this](https://github.com/helix-editor/helix/pull/2608/commits/cc19db1e5306e6cbb396ae33c51e99927e879584#diff-3626b17eb15f4783dd83a9f709ecef3bd3f591db4b22d83ee2603ff0a09d49e5) before that micro-optimization (I wasn't sure about it anyway, as it just saves 1 clone...)

Philipp-M replied:
Yeah an `impl PartialEq<CompletionItem> for LspCompletionItem` is much better than that macro, for whatever reason, I haven't thought of that first...

I've reverted to that previous change, and added `PartialEq` impls.

the-mikedavis commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +301,226 @@ fn request_completion(
+            .filter_map(Result::ok)
+            .filter_map(|dir_entry| {
+                let path = dir_entry.path();
+                // check if <chars> in <path>/<chars><cursor> matches the start of the filename
+                let filename_starts_with_prefix =
+                    match (path.file_name().and_then(OsStr::to_str), &typed_file_name) {
+                        (Some(re_stem), Some(t)) => re_stem.starts_with(t),
+                        _ => true,
+                    };
+                if filename_starts_with_prefix {
+                    dir_entry.metadata().ok().map(|md| (path, md))
+                } else {
+                    None
+                }
```
I believe we want to skip the prefix check here and delegate the filtering to the ui::Menu and Nucleo. This makes a difference if you trigger completion manually, for example if you're in the helix repo and type `./lsp` you'll see `helix-lsp` and `helix-lsp-types`. If you then hit `C-c` and `C-x` to manually trigger completion though you won't have any suggestions because of the prefix check.

So instead this could be `.filter_map(|dir_entry| dir_entry.metadata().ok().map(|md| (dir_entry.path(), md)))`

Philipp-M replied:
Yeah makes sense, it also allows case-insensitive completion I think.

kik4444 commented:
Not sure if anyone's mentioned this before, but have you considered taking inspiration or logic from the [nvim cmp-path plugin](https://github.com/hrsh7th/cmp-path)? As far as I've noticed it doesn't attempt any complexities like Home-path resolution except the tilde `~`.

Philipp-M commented:
I have not, but since it's implemented already (and it's not really complex IMO), I guess we can keep support for resolution of `$HOME` and `${HOME}`?

kik4444 commented:
Well if it works it works. Either way I'd just be grateful to have this as part of the main editor. This feature is on the list of things I'd like to have before giving Helix a more serious attempt.

nrdxp commented:
> Well if it works it works. Either way I'd just be grateful to have this as part of the main editor. This feature is on the list of things I'd like to have before giving Helix a more serious attempt.

fwiw I have been using Helix for years now and just carrying this along as a patch pretty much that whole time. It works and I've never had an issue with it. Hope it get's merged soon

RGBCube commented:
Seems like all reviews are passing and the code works, maybe it's time to finally merge this ~2 year old MR?

David-Else commented:
I am using it and have had no problems, could not live without it now.

the-mikedavis commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+                    .to_string_lossy()
+                    .to_string();
+
+                let full_path = path.canonicalize().unwrap_or_default();
+                let full_path_name = full_path.to_string_lossy().into_owned();
```
For canonicalization we generally prefer `helix_stdx::path::canonicalize` now (it doesn't expand symlinks like the canonicalize in `std`). For the full path name that we display in the documentation let's pass that canonicalized path through `helix_stdx::path::fold_home_dir` so it displays better

Philipp-M replied:
Hmm, I wonder though, if a link *should* be shown resolved then or not (because then I think we still need `std::path::canonicalize`), which IMO makes sense but I have no strong opinion here...

the-mikedavis replied:
I think we should try to avoid expanding symlinks in the parent directories as std's canonicalize would do. For an edge case, my `~/src/` directory is a symlink to `/nix/persist/home/michael/src` so canonicalizing a path like `./helix-event` would expand all the way to `/nix/persist/home/michael/src/helix/hx/helix-event`. If the directory entry that we are looking at is a symlink though we should probably follow that link (once?).

Philipp-M replied:
Yeah it's what I'd expect, but I see your point.

I think following a link once, would result in otherwise inconsistent behavior but I'm not sure. So I'd rather not follow it, and display it as link then, I've pushed a version without following the link for now.

the-mikedavis reviewed (commented):
I only have some minor style nits and a comment about canonicalization and display, otherwise I think this looks good :+1:

pascalkuthe commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+
+    let (dir_path, typed_file_name) = PATH_REGEX
+        .find_iter(line_until_cursor)
+        .last() // this is a workaround for missing "match end of slice" support in `rope::Regex`
```
regex cursor works the same way as regex automata with regards to anchors/search ranges. What you need to do is anchor you regex (put `$` at the end) and instead of using `text.slice(...).regex_input_at(start..cursor` us `text.slice(start..cursor).regex_input_at(..)`. 

The difference is that `regex_input_at(start..end)` runs a regex within a part of a string/rope but still takes the surrounding context into account for lookaround assertions (so `\b` works correctly and so `^` only matches start/end of entire string and not the start of the search range).

pascalkuthe commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+                return None;
+            }
+            let start = text.byte_to_char(matched_path.start());
+            let matched_path = &text.slice(start..end).to_string();
```
nit: just do `text.byte_slice(matched_pathn.start()..matched_path.end())` instead of converting back and fourth a bunch of times

Philipp-M commented:
Thanks for the review, I think I handled all the feedback.

the-mikedavis commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +302,197 @@ fn request_completion(
+    let line_until_cursor = text.slice(start..cursor).regex_input_at(..);
+
+    let (dir_path, typed_file_name) = PATH_REGEX.search(line_until_cursor).and_then(|m| {
+        let matched_path = &text.byte_slice(m.start()..m.end()).to_string();
```
Now that we `slice` and then use `regex_input_at` we don't have the byte indices automatically adjusted for us. (They're relative to the slice `text.slice(start..cursor)` - `regex_input_at` made them absolute to `text` previously.)

To adjust this we'll need to grab the start byte (probably easiest to do `let start_byte = text.char_to_byte(start);` after `start` is bound) and then our matched path would be `text.byte_slice(start_byte + m.start()..start_byte + m.end())`

Philipp-M replied:
Whoops, I guess I wasn't thorough enough... thanks.

pascalkuthe commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+                    ""
+                };
+
+                let documentation = {
```
this should also just be a seperate function

pascalkuthe commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+            .filter_map(Result::ok)
+            .filter_map(|dir_entry| dir_entry.metadata().ok().map(|md| (dir_entry.path(), md)))
+            .map(|(path, md)| {
+                let file_name = path
```
this is a bit ugly, what `dir_entry.path()` does is just `dir_path.join(dir_entry.file_name())` so going trough all this to extractd the file name again doesn't seem ideal. Instead just extract the file name and use that here (and just do `dir_path.join(file_name)` below)

Philipp-M replied:
Should be handled in the recent push.

pascalkuthe commented on helix-view/src/document.rs:
```diff
@@ -1713,6 +1713,12 @@ impl Document {
         self.version
     }
 
+    pub fn supports_path_completion(&self) -> bool {
```
this seem like the wrong name, it should be `path_completions_enabled()` (or similar)

pascalkuthe commented on helix-term/src/ui/completion.rs:
```diff
@@ -82,18 +87,101 @@ impl menu::Item for CompletionItem {
+    Path(PathCompletionItem),
+}
+
+impl PartialEq<CompletionItem> for LspCompletionItem {
```
I don't think we necessarily need these

Philipp-M replied:
See https://github.com/helix-editor/helix/pull/2608#discussion_r1740929468

pascalkuthe commented on helix-term/src/ui/completion.rs:
```diff
@@ -82,18 +87,101 @@ impl menu::Item for CompletionItem {
+#[derive(Debug, PartialEq, Clone)]
+pub enum CompletionItem {
+    Lsp(LspCompletionItem),
+    Path(PathCompletionItem),
```
I don't think we need a seperate `PathCompletionItem` type. Instead we can just have a `kind: Cow<str>`  on `helix_core::CompletionItem`. At that point we could also just call that variant `Other` (so it's clear there will only ever be an lsp variant and one for everything else going trough the helix-core type).

I am also not really sure about this abstraction in general. I think first starters this should be much better encapsulated. This file is certainly the wrong place for it (I don't want the handler/resolve code to depend on the UI at all). So probably at least a submodule in `handlers::completion`. I would also like to have as much information extraction as possible within methods (like `preselect`). It's ok to keep `lsp_item_to_transaction in the UI code for now since that is a larger change (altough it also doesn't belong there)

Philipp-M replied:
> I don't think we need a seperate `PathCompletionItem` type. Instead we can just have a kind: `Cow<str>` on `helix_core::CompletionItem`. At that point we could also just call that variant `Other` (so it's clear there will only ever be an lsp variant and one for everything else going trough the helix-core type).

I just fear that there may be more information relevant for a completion item, but that fear may well be unjustified.

> This file is certainly the wrong place for it

See [recent discussion](https://github.com/helix-editor/helix/pull/2608#discussion_r1720817397) for it, I'm aware of it, but I don't want to start an even bigger refactor, when PRs are open for so long without any kind of review to avoid resolving merge-conflicts. It shouldn't be too hard to move this somewhere else, when that kind of refactor is done.

> At that point we could also just call that variant `Other`.

pascalkuthe replied:
> I just fear that there may be more information relevant for a completion item, but that fear may well be unjustified.

To me if there is more relevant information is should be added to the generic `helix_core::CompletionItem`. I don't want to couple the UI code to the completion sources. Basically `helix_core::CompletionItem` shouldn't be modeled after whatever data completion sources can or can't provide but instead by the data that the UI can actually support (and I would want everything to go trough that generic interface except LSP). The only reason that lsp needs to be separate is because creating the transaction is complicated and needs to be done lazily.

pascalkuthe replied:
> See https://github.com/helix-editor/helix/pull/2608#discussion_r1720817397 for it, I'm aware of it, but I don't want to start an even bigger refactor, when PRs are open for so long without any kind of review to avoid resolving merge-conflicts. It shouldn't be too hard to move this somewhere else, when that kind of refactor is done.

I agree that we shouldn't do a large refactor in this PR but moving this into another file inside helix term (like in the handler) is a pretty easy cosmetic change and reduces spaghetti code by a lot. I don't think that would even conflict with the snippet system PR

Philipp-M replied:
I think this should be handled.

Philipp-M commented on book/src/editor.md:
```diff
@@ -32,6 +32,7 @@
 | `cursorcolumn` | Highlight all columns with a cursor | `false` |
 | `gutters` | Gutters to display: Available are `diagnostics` and `diff` and `line-numbers` and `spacer`, note that `diagnostics` also includes other features like breakpoints, 1-width padding will be inserted if gutters is non-empty | `["diagnostics", "spacer", "line-numbers", "spacer", "diff"]` |
 | `auto-completion` | Enable automatic pop up of auto-completion | `true` |
+| `path-completion` | Enable filepath completion. Show files and directories if an existing path at the cursor was recognized, either absolute or relative to the current opened document or current working directory (if the buffer is not yet saved). Defaults to true. | `true` |
```
Maybe this makes sense to add, to avoid questions, why it's not working in windows?
```suggestion
| `path-completion` | Enable filepath completion. Show files and directories if an existing path at the cursor was recognized, either absolute or relative to the current opened document or current working directory (if the buffer is not yet saved). This is currently only supported in Linux and OSX. Defaults to true. | `true` |
```

Probably makes sense as well to open an issue, or update #2874

pascalkuthe replied:
I think the code does work on windows (or at least after you use the regex I suggested below)

pascalkuthe commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +302,200 @@ fn request_completion(
+            static HOME_DIR: Lazy<Option<OsString>> = Lazy::new(|| std::env::var_os("HOME"));
+
+            let home_resolved_path = if let Some(home) = &*HOME_DIR {
+                let first_separator_after_home = if matched_path.starts_with("~/") {
```
I think for this feature in particular (but also gf) proper path expansion would be nice instead of this special case.

Philipp-M commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+                        Some(1)
+                    } else if matched_path.starts_with("$HOME") {
+                        Some(5)
+                    } else if matched_path.starts_with("${HOME}") {
```
You mean implement our own (more general) kind of shell expansion in `helix-stdx`? What kind of dependency are you talking about?

pascalkuthe replied:
I meant https://docs.rs/shellexpand/latest/shellexpand/

A lot of complexity in this crate comes from the fact that in the past converting raw OsString <-> bytes was ugly but these days it's almost trivial (since we can use a regex).

Basically just iterate all matches for a regex `\$\{[\w@/\\.\-+#$%~=:]+\}||\$\w+` and try to replace them with the environment variable. I had a function written up. If you feel uncomfortable with that I can also write it

Philipp-M replied:
I mean I could certainly dive into this, but it seems you already have a pretty good idea how to do all of that, so you're likely faster (including much more research regarding the path recognition regex).

Maybe it makes sense for you to create a PR with `gf` + path regex + shell expansion, and I base this path-completion PR on top of that?

I can help reviewing, if that helps getting this merged faster.

pascalkuthe replied:
I would prefer doing a PR against your branch/fork

Philipp-M replied:
That's good for me as well.

Philipp-M replied:
This was solved with https://github.com/Philipp-M/helix/pull/4

Philipp-M commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+        return None;
+    }
+
+    // TODO find a good regex for most use cases (especially Windows, which is not yet covered...)
```
Nice one, I'll take a look at it.

Philipp-M replied:
Should be handled by https://github.com/Philipp-M/helix/pull/4

Philipp-M commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+                _ => false,
+            };
+            // check if there are chars after the last slash, and if these chars represent a directory
+            match std::fs::metadata(path.clone()).ok() {
```
It could well be that this is an artifact, after a [recent change](https://github.com/helix-editor/helix/pull/2608#discussion_r1740152557), I've overlooked, I'll take a look into it.

Philipp-M replied:
I removed that, is the current version better?

pascalkuthe commented:
pushed to this branch by accident, sorry about that I reverted the change (not clear in github UI)

Philipp-M commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+            let matched_path = &text.slice(start..end).to_string();
+
+            // resolve home dir (~/, $HOME/, ${HOME}/) on unix
+            #[cfg(unix)]
```
Should be handled by https://github.com/Philipp-M/helix/pull/4

Philipp-M commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
     });
 }
 
+fn path_completion(
```
I've moved this into `handlers/completion/path.rs`

Philipp-M commented on helix-term/src/handlers/completion.rs:
```diff
@@ -291,6 +300,213 @@ fn request_completion(
+                }
+            }
+            let ends_with_slash = match matched_path.chars().last() {
+                Some(std::path::MAIN_SEPARATOR) => true,
```
Should be handled by https://github.com/Philipp-M/helix/pull/4

Philipp-M commented on helix-term/src/handlers/completion.rs:
```diff
@@ -251,15 +256,19 @@ fn request_completion(
             }
+            .boxed()
         })
+        .chain(path_completion(cursor, text.clone(), doc))
```
I've added an `Arc<AtomicBool>` for this which is polled.
Alterntative would probably just be using tokio async functions, but as I said, it was significantly slower (probably around half the speed) than using everything in a blocking call.

Philipp-M commented:
AFAICS path detection + shell-expansion should now be quite a bit more robust with @pascalkuthe s changes, and should also support windows.
I have no (easy) way currently test this though, so real-world testing would be appreciated.

pascalkuthe commented on helix-term/src/handlers/completion.rs:
```diff
@@ -196,6 +202,7 @@ fn request_completion(
     // necessary from our side too.
     trigger.pos = cursor;
     let trigger_text = text.slice(..cursor);
+    let cancel_completion = Arc::new(AtomicBool::new(false));
```
I don't want to build custom synchronization code here around atomic. What you did is reasonable but I want it to be part of a single cancelation token implementation. I thought our cancelation token was already clonable when I made that comment but seems I misremembered.

I already have a specific implementation in mind (a simplified variant if the tokio-utils cancelation token). I am currently on vacation but I will send a PR for that to your branch once I get back (unless you want to /get around to implementing that first)

Philipp-M replied:
I can wait for you PR, thanks.

pascalkuthe replied:
I put up a PR, ended up a bit more complex than I thaught but I am very happy with it. After looking at it I also did a lot of "quick hacks" like this that weren't really wrong but are better centralized in dedicated shared code

Philipp-M replied:
Yeah, probably better to have this a little bit more encapsulated.

pascalkuthe commented on helix-term/src/handlers/completion.rs:
```diff
@@ -346,7 +368,14 @@ pub fn trigger_auto_completion(
+    let trigger_path_completion = matches!(
+        text.get_bytes_at(text.len_bytes())
+            .and_then(|t| t.reversed().next()),
+        Some(b'/' | b'\\')
```
I think we only want to match the backslash on windows

RGBCube replied:
I don't think this should be platform specific. One could be writing code for mainly-Windows software (like PowerShell) while being on a Linux machine.

Philipp-M replied:
> I think we only want to match the backslash on windows

Yeah I thought about doing this, but seemingly forgot about it not long after...

> One could be writing code for mainly-Windows software (like PowerShell) while being on a Linux machine

But on a linux machine there's no `\` as path separator is there? So it would not make sense to have this on Linux (as this path needs to exist and be readable for path-completion to work)?

RGBCube replied:
Treating `/` as `\` while serving completions on Linux is fine as Windows does the reverse, there's no need to complicate behavior for no benefit.

I'm sure someone out there is editing Windows specific source code on Linux, where they treat `foo/bar` as `foo\bar`

pascalkuthe replied:
Windows path syntax and Linux path syntax are not generally equivalent. The path detection is also OS dependent, backslashes act as escape chars (like with spaces) on unix but oath separators on windows.

Windows enables forward slashes as path separators in their paths so we support that but on unix you can absolutely not treat backslash as separators.

Some Windows paths also just don't make sense on unix. If you want crossplatform path completions always use forward slash since they work both on windows and unix (including powershell)

Philipp-M replied:
Should be implemented now.

RGBCube replied:
Fair, I forgot that stuff like `cat foo\ bar` was valid posix shell for `cat "foo bar"`. I didn't know that `/` was that portable.

pascalkuthe commented on helix-term/src/handlers/completion/path.rs:
```diff
@@ -0,0 +1,194 @@
+                Some(parent_dir) if path.is_relative() => parent_dir.join(&path),
+                _ => path.into_owned(),
+            };
+            let ends_with_slash = matches!(matched_path.as_bytes().last(), Some(b'/' | b'\\'));
```
This may have been my mistake but I think we want to only set ends with slash to true for backslash if we are on windows (so for windows we want to check for both forward and reverse slash and for unix we only want to check for forward)

pascalkuthe commented on helix-term/src/handlers/completion/path.rs:
```diff
@@ -0,0 +1,194 @@
+                    return None;
+                }
+
+                let file_name_str = file_name.to_string_lossy().to_string();
```
This allocated, I think we can't properly compelte non-utf8 filename so let's just skip them (so us into_string() and ignore cases that return none)

pascalkuthe reviewed (commented):
This is starting to look pretty good, thanks for addressing my comments! I found a couple smaller things but hopefully should be ready soon

David-Else commented:
When using the `%` macro with this branch the home directory is not included: https://github.com/helix-editor/helix/discussions/12002.

Check this out with a file in your Linux home directory:

```toml
[keys.normal.space]
o = "@:o <C-r>%<C-w>"
```

Philipp-M commented:
I'm not sure what the expected outcome is, or whether this is a bug at all?

I've compared the `path-completion` branch with current `master` and it seems to have the same behavior for me.
I don't think there should any difference to `master` for this macro reasoning from the code as the new shell-expansion doesn't seem to be used in the `open` impl.

Nice macro btw.

David-Else commented:
@Philipp-M I don't think it is a problem or a bug, chatting about it in the linked discussion.

the-mikedavis reviewed (approved):
First thanks for working on this @Philipp-M and especially for keeping the branch alive for so long[^1]! 😀

I've been running this locally for a few weeks now and I'm pretty sure it's good to go. I'd like to merge it down and have it sit in master for a while to eek out any bugs we might've introduced in the completion changes. I might've felt a change in the ranges we replace with a completion, but this might be a rust-analyzer peculiarity or me imagining things so I don't want it to block the merge. I'd like to see more eyes from anyone running master before a possible December release.

Using this in some Nix files and, nicely, `Cargo.toml`s has been quite nice. Thanks again @Philipp-M! (And completion and regex improvements from @pascalkuthe ofc :)

This is one small step for non-LSP completions and one giant leap for moving the codebase away from its very close relationship with LSP :rocket:

[^1]: Some of the other PRs merged recently have been 10,000 + this PR's number 😅

RoloEdits commented:
@the-mikedavis I have also been experiencing some issues with `rust-analyzer`, without this branch's changes, which I expanded on more in #11933. So shouldn't have anything to do with this pr.

RoloEdits commented:
I did notice that the msrv was broken though. Some of the `OsStr` stuff needs rust 1.74.0

```shell
warning: current MSRV (Minimum Supported Rust Version) is `1.70.0` but this item is stable since `1.74.0`
  --> helix-stdx\src\env.rs:98:21
   |
98 |     let bytes = src.as_encoded_bytes();
   |                     ^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incompatible_msrv
   = note: `#[warn(clippy::incompatible_msrv)]` on by default

warning: current MSRV (Minimum Supported Rust Version) is `1.70.0` but this item is stable since `1.74.0`
   --> helix-stdx\src\env.rs:117:28
    |
117 |         let var = unsafe { OsStr::from_encoded_bytes_unchecked(var) };
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incompatible_msrv

warning: current MSRV (Minimum Supported Rust Version) is `1.70.0` but this item is stable since `1.74.0`
   --> helix-stdx\src\env.rs:124:25
    |
124 |                     val.as_encoded_bytes()
    |                         ^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incompatible_msrv

warning: current MSRV (Minimum Supported Rust Version) is `1.70.0` but this item is stable since `1.74.0`
   --> helix-stdx\src\env.rs:138:18
    |
138 |         unsafe { OsString::from_encoded_bytes_unchecked(res) }.into()
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incompatible_msrv
```

poliorcetics commented:
IIRC Helix uses Firefox's MRSV ? It's currently 1.76.0 (https://firefox-source-docs.mozilla.org/writing-rust-code/update-policy.html) so we could just increase it
